src/main/java/uk/gov/hmcts/ccd/domain/service/common/CaseAccessGroupUtils.java

```diff
@@ -0,0 +1,236 @@
+    protected static final String ORG_IDENTIFIER_TEMPLATE = "$ORGID$";
+    protected static final String ORG_POLICY_CASE_ASSIGNED_ROLE = "OrgPolicyCaseAssignedRole";
+    protected static final String CASE_ACCESS_GROUP_TYPE = "caseAccessGroupType";
+    protected static final String GROUPACCESS_VALUE = "value";
+    protected static final String GROUPACCESS_ID = "id";
+
+    private CaseDataService caseDataService;
+    private CaseTypeDefinition caseTypeDefinition;
+
+    protected static final String GROUPACCESS_CLASSIFICATION = "classification";
+
+    public void updateCaseAccessGroupsInCaseDetails(CaseDetails caseDetails,
+                                                    CaseTypeDefinition caseTypeDefinition,
+                                                    CaseDataService caseDataService) {
+        this.caseDataService = caseDataService;
+        this.caseTypeDefinition = caseTypeDefinition;
+
+        removeCCDAllCasesAccessFromCaseAccessGroups(caseDetails);
+
+        List<AccessTypeRoleDefinition> accessTypeRoleDefinitions =
+            caseTypeDefinition.getAccessTypeRoleDefinitions();
+        List<AccessTypeRoleDefinition> filteredAccessTypeRolesDefinitions = filterAccessRoles(caseDetails,
+            accessTypeRoleDefinitions);
+        if (filteredAccessTypeRolesDefinitions != null) {
+            List<CaseAccessGroupWithId> caseAccessGroupWithIds = new ArrayList<>();
+            for (AccessTypeRoleDefinition acd : filteredAccessTypeRolesDefinitions) {
+                JsonNode caseAssignedRoleFieldNode = findOrganisationPolicyNodeForCaseRole(caseDetails,
+                    acd.getCaseAssignedRoleField());
+                if (caseAssignedRoleFieldNode != null) {
+                    String orgIdentifier = caseAssignedRoleFieldNode.get(ORGANISATION)
+                        .get(ORGANISATIONID).textValue();
+
+                    if (orgIdentifier != null && !orgIdentifier.isEmpty()) {
+                        String caseGroupID = acd.getCaseAccessGroupIdTemplate()
+                            .replace(ORG_IDENTIFIER_TEMPLATE, orgIdentifier);
+
+                        CaseAccessGroup caseAccessGroup = CaseAccessGroup.builder().caseAccessGroupId(caseGroupID)
+                            .caseAccessGroupType(CCD_ALL_CASES).build();
+                        CaseAccessGroupWithId caseAccessGroupWithId = CaseAccessGroupWithId.builder()
+                            .caseAccessGroup(caseAccessGroup).id(UUID.randomUUID().toString()).build();
+
+                        caseAccessGroupWithIds.add(caseAccessGroupWithId);
+                    }
+                }
+            }
+            setUpModifiedCaseAccessGroups(caseDetails, caseAccessGroupWithIds);
+        }
+
+    }
+
+    private void setUpModifiedCaseAccessGroups(CaseDetails caseDetails,
+                                               List<CaseAccessGroupWithId> caseAccessGroupWithIds) {
+        ObjectMapper mapper = new ObjectMapper();
```

Comment:
Either inject this (I believe there's already a bean defined at the app level), or create at class level, rather than creating a new instance every time.

Reply:
fixed in #2389